### PR TITLE
PR: Allow diff & shift to take lists of periods? #19298

### LINKED ISSUE #19298: Allow diff & shift to take lists of periods?
I've seen code like

```python
In [36]: s = pd.Series(range(10))

In [37]: pd.concat([s.diff(i).rename(f"L{i}") for i in range(10)], axis=1)
Out[37]:
    L0   L1   L2   L3   L4   L5   L6   L7   L8   L9
0  0.0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
1  0.0  1.0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
2  0.0  1.0  2.0  NaN  NaN  NaN  NaN  NaN  NaN  NaN
3  0.0  1.0  2.0  3.0  NaN  NaN  NaN  NaN  NaN  NaN
4  0.0  1.0  2.0  3.0  4.0  NaN  NaN  NaN  NaN  NaN
5  0.0  1.0  2.0  3.0  4.0  5.0  NaN  NaN  NaN  NaN
6  0.0  1.0  2.0  3.0  4.0  5.0  6.0  NaN  NaN  NaN
7  0.0  1.0  2.0  3.0  4.0  5.0  6.0  7.0  NaN  NaN
8  0.0  1.0  2.0  3.0  4.0  5.0  6.0  7.0  8.0  NaN
9  0.0  1.0  2.0  3.0  4.0  5.0  6.0  7.0  8.0  9.0
```

Could we simplify that by allowing `s.diff(range(10), prefix="L")`?

How would we handle DataFrames? Prefix with the column name and a separator, like `get_dummies`?

### PR DESCRIPTION
**Pull request by tliu30**
- [x] implement for Series.diff and DataFrame.diff
- [ ] implement for Series.shift and DataFrame.shift
- [ ] closes #19298
- [ ] tests added / passed
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

**Review by jreback — CHANGES_REQUESTED** [review]
pls make a test, not really sure about the impl

File: pandas/core/series.py [hunk 1]
```diff
@@ -2002,8 +2002,25 @@ def diff(self, periods=1):
...
         """
-        result = algorithms.diff(com._values_from_object(self), periods)
-        return self._constructor(result, index=self.index).__finalize__(self)
+        def _diff(x):
+            result = algorithms.diff(com._values_from_object(self), x)
+            return self._constructor(result, index=self.index).__finalize__(self)
+
+        if isinstance(periods, int):
+            return _diff(periods)
+        elif isinstance(periods, list):
```
Comment (TomAugspurger) [inline pandas/core/series.py]: Shouldn't hardcode `list` here. Any sequence is fine.

File: pandas/core/series.py [hunk 2]
```diff
@@ -2002,8 +2002,25 @@ def diff(self, periods=1):
...
+            if len(periods) == 0:
+                raise ValueError('Must provide non empty list to periods')
+            else:
+                result = _diff(periods[0]).to_frame().add_suffix('{}{}'.format(suffix, periods[0]))
+                for period in periods[1:]:
+                    next_addition = _diff(period).to_frame().add_suffix('{}{}'.format(suffix, period))
+                    result = result.join(next_addition)
```
Comment (TomAugspurger) [inline pandas/core/series.py]: This is doing `len(periods)` joins. Better to build up a list of frames and then concat.

**Review by TomAugspurger — COMMENTED** [review]
Haven't looked closely at the  implementation yet either.

Can you see if `algorithms.diff` can be expanded to take a list of integers for `n`? That'd be faster than calling `diff` repeatedly.

File: pandas/core/series.py (hunk 2 above)
Comment (tliu30) [reply pandas/core/series.py]: I didn't want to call concat b/c it raised the possibility of circular imports (I *think* the file in which pd.concat is defined also imports pd.Series)

**Comment by tliu30** [comment]
thanks tom & jeff

@TomAugspurger will take a look at algorithms.diff; any thoughts on `pd.DataFrame(periods=[...], axis=1)`? instinct right now is to just not allow multiple periods to be passed when `axis=1`

**Comment by TomAugspurger** [comment]
I think raising in that case is fine for now.

You can import concat inside the `diff` method. We do this elsewhere.

**Comment by jreback** [comment]
closing as stale, if you want to continue working, pls ping and we can re-open. you will need to merge master.

**Pull request closed without merging**